Refactor logs using structlog

bigspider opened the pull request:
By using [structlog](https://www.structlog.org/en/stable/), it's easier to keep the logs structured and decide how to render them for different purposes.

This PR rewrites the logging module to use structlog, keeping them for now as plaintext, with the following settings:
- the logs to console is at INFO level
- the logs to file is at DEBUG level

Also added a "daemon" parameter, that suppresses output to console. It's not a proper daemon support, as the shell is not released; that's not trivial to do and is left for separate work once we figure out the best way to do it.

This should make it relatively easy to integrate any log aggregator, like the kibana dashboard after #148 is merged.

sr-gi commented on teos/__init__.py:
```diff
@@ -17,6 +16,7 @@
     "BTC_FEED_PROTOCOL": {"value": "tcp", "type": str},
     "BTC_FEED_CONNECT": {"value": "localhost", "type": str},
     "BTC_FEED_PORT": {"value": 28332, "type": int},
+    "DAEMON": {"value": 0, "type": int},
```
Actually, read comment in teosd.

bigspider replied:
Dropped the two commits adding the daemon config.

sr-gi commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-             msg (:obj:`str`): the message to be logged.
-             kwargs (:obj:`dict`): a ``key:value`` collection parameters to be added to the output.
-        """
+    """
```
Also some nits: spance between parameter and type (`param (:obj: ...)`)

No need of parenthesis fort the Return and Raises

bigspider replied:
Fixed.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
+    global configured
+
+    if configured:
+        raise RuntimeError("logging was already configured.")
```
Well, I think calling this twice means that your code is broken, so it's good if the exception is not handled. Silently failing (and returning None that will likely be ignored) might lead to not detecting that you called it twice - and maybe running with a config you don't expect.

sr-gi replied:
Well, I don't think the latter would apply, since the the first time it will configure it with the original conf and the second time will not run it at all, so it would still run with the original conf.

I don't think this is a big deal though, since there is no chance this is called more than once.

If you rather leaving it like this: Cap "Logging" and remove the tailing dot

bigspider replied:
Well, that mean we are choosing the first config, but we don't know which one was the intended one if two are given.

Updated error message in 56adb9b.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
 import json
 import logging
+import logging.config
```
Removed all.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-    Args:
-        log_name_prefix (:obj:`str`): the prefix of the logger where the data will be stored in (server, client, ...).
-        actor (:obj:`str`): the system actor that is logging the event (e.g. ``Watcher``, ``Cryptographer``, ...).
+    Render ``event_dict``. It renders the timestamp, followed by the actor within "[]" (unless it's None),
```
Fixed.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-        return msg if self.actor is None else "[{}]: {}".format(self.actor, msg)
+    def _repr(self, val):
+        """
+        Determine representation of *val* depending on its type.
```
Changed to a one-line description, and also made the code more compact with a ternary if/else.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-        log_name_prefix (:obj:`str`): the prefix of the logger where the data will be stored in (server, client, ...).
-        actor (:obj:`str`): the system actor that is logging the event (e.g. ``Watcher``, ``Cryptographer``, ...).
+    Render ``event_dict``. It renders the timestamp, followed by the actor within "[]" (unless it's None),
+    followed by the event, then any remaining item in event_dict in the key=value format
```
Added.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-            # Remove the extra 2 characters (space and comma) and add all data to the final message.
-            message += " ({})".format(params[:-2])
+        actor = event_dict.pop("actor", None)
+        if actor is not None:
```
Done for both `ts` and `actor` (now called `component`).

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-    def _create_console_message(self, msg, **kwargs):
-        s_message = _StructuredMessage(self._add_prefix(msg), **kwargs).to_dict()
-        message = "{} {}".format(s_message["time"], s_message["message"])
+    def __call__(self, _, __, event_dict):
```
Added a one-liner docstring, not sure there's more to add as it would repeat the class docstring, basically.

bigspider commented on teos/teosd.py:
```diff
@@ -53,7 +53,15 @@ def main(command_line_conf):
         setup_data_folder(data_dir)
-        setup_logging(config.get("LOG_FILE"), LOG_PREFIX)
+
+        daemon = config.get("DAEMON") != 0
```
Dropped the two commits adding the daemon config.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
-             msg (:obj:`str`): the message to be logged.
-             kwargs (:obj:`dict`): a ``key:value`` collection parameters to be added to the output.
-        """
+    """
```
Hard to say exactly what's the return value, as it's inherited from structlog. Added a short description without the type.

bigspider commented on common/logger.py:
```diff
@@ -1,96 +1,133 @@
+        # force event to str for compatibility with standard library
+        event = event_dict.pop("event")
+        if not isinstance(event, str):
+            event = str(event)
```
Changed using `_repr` now.

bigspider commented:
Should have addressed almost everything, still a comment suspended about what happens if `setup_logging` is called twice.

I added tests for the expected output of CustomLogRenderer in various combinations; not sure what tests to add for setup_logging and get_logger. For get_logger, the problem is also that the object returned is kinda opaque: a test that would make sense is to check whether the `component` is indeed bound to the logger, but didn't find a way without "cheating" by reading private variables in the returned object.

sr-gi commented on common/logger.py:
```diff
@@ -1,96 +1,124 @@
-    @staticmethod
-    def _create_file_message(msg, **kwargs):
-        return json.dumps(_StructuredMessage(msg, **kwargs).to_dict())
+        # Represent all the key=value elements still in event_dict
```
We can surround this by parenthesis so it is easier to parse in case we ever need to:

ts [component] msg (key1=value1, ...)

bigspider replied:
Good idea, done in e2601494837b42927ed1b1a4afa933c6a5336f0d.

bigspider commented:
Should have addressed all the comments. Added the `logger` to the docstrings (without specifying the type, though). Added a test for `get_logger` which is a bit unusual, but the only way I found to test it using only the public interface of the object returned by `get_logger`. There are still no tests for `setup_logger`, but I don't thing there's much to test there, tbh.
